src/Api/ShareaholicApi.php

```diff
@@ -0,0 +1,171 @@
+  const API_URL = 'https://web.shareaholic.com';
+  const HEALTH_CHECK_URL = self::API_URL . '/haproxy_health_check';
+  const KEY_GENERATING_URL = self::API_URL . '/publisher_tools/anonymous';
+  const EVENTS_URL = self::API_URL . '/api/events';
+  const SESSIONS_URL = self::API_URL . '/api/v3/sessions';
+
+  /** @var Client */
+  private $httpClient;
+
+  /** @var ImmutableConfig */
+  private $config;
+
+  /** @var LoggerInterface */
+  private $logger;
+
+  public function __construct(Client $httpClient, ImmutableConfig $config, LoggerInterface $logger) {
+    $this->httpClient = $httpClient;
+    $this->config = $config;
+    $this->logger = $logger;
+  }
+
+  /**
+   * @return string|null
+   */
+  public function generateApiKey($verificationKey, $siteName, $langcode) {
+
+    $post_data = [
+      'configuration_publisher' => [
+        'verification_key' => $verificationKey,
+        'site_name' => $siteName,
+        'domain' => \Drupal::request()->getHost(),
+        'platform_id' => '2',
+        'language_id' => $this->getLanguageId($langcode),
+        'shortener' => 'shrlc',
+        'recommendations_attributes' => [
+          'locations_attributes' => [],
+        ],
+        'share_buttons_attributes' => [
+          'locations_attributes' => [],
+        ],
+      ],
+    ];
+
+    $client = $this->httpClient;
+    $apiUrl = static::KEY_GENERATING_URL;
+    $settings = [
+      'headers' => [
+        'Content-Type' => 'application/vnd.api+json',
+        'User-Agent' => $this->getUserAgent(),
+      ],
+      'body' => json_encode($post_data),
+    ];
+
+    try {
+      $response = $client->post($apiUrl, $settings);
+      $data = (string) $response->getBody();
+
+      if (empty($data)) {
+        return NULL;
+      }
+
+    } catch (RequestException $e) {
+      return NULL;
+    }
+
+    $json_response = json_decode($data, TRUE);
+    return $json_response['api_key'];
+  }
+
+  /**
+   * @return string|null
+   */
+  public function getPublisherToken() {
+    $response = $this->httpClient->post(self::SESSIONS_URL, [
+      RequestOptions::JSON => [
+        'site_id' => $this->config->get('api_key'),
+        'verification_key' => $this->config->get('verification_key'),
+      ],
+    ]);
+
+    $statusCode = $response->getStatusCode();
+
+    if (preg_match('/20*/', $statusCode)) {
+      $body = json_decode($response->getBody()->getContents(), TRUE);
+      if (!empty($body['publisher_token'])) {
+        return $body['publisher_token'];
+      }
+      $this->logger->critical("Publisher token couldn't be received. Wrong content of server's response.");
+    } else {
+      $this->logger->critical("Publisher token couldn't be received. Request status code: $statusCode");
+    }
+
+    return NULL;
+  }
+
+  /**
+   * @return bool
+   */
+  public function connectivityCheck() {
+    $health_check_url = self::HEALTH_CHECK_URL;
+    $response = $this->httpClient->get($health_check_url);
```

Comment:
What would happen if Drupal doesn't have access to the internet?

Reply:
Good catch. There would be an exception. Fixed.